pcsx2/USB/usb-pad/usb-train.cpp

```diff
@@ -0,0 +1,546 @@
+		{
+			case TRAIN_TYPE2:
+				out.type2.control = 0x1;
+				out.type2.brake = s->passthrough ? s->data.brake : dct01_brake(s->data.brake);
+				out.type2.power = s->passthrough ? s->data.power : dct01_power(s->data.power);
+				out.type2.horn = 0xFF; // Button C doubles as horn.
+				out.type2.hat = s->data.hatswitch;
+				out.type2.buttons = dct01_buttons(s->data.buttons);
+				break;
+			case TRAIN_SHINKANSEN:
+				out.shinkansen.brake = s->passthrough ? s->data.brake : dct02_brake(s->data.brake);
+				out.shinkansen.power = s->passthrough ? s->data.power : dct02_power(s->data.power);
+				out.shinkansen.horn = 0xFF; // Button C doubles as horn, skip.
+				out.shinkansen.hat = s->data.hatswitch;
+				out.shinkansen.buttons = dct02_buttons(s->data.buttons);
+				break;
+			case TRAIN_RYOJOUHEN:
+				out.ryojouhen.brake = s->passthrough ? s->data.brake : dct03_brake(s->data.brake);
+				out.ryojouhen.power = s->passthrough ? s->data.power : dct03_power(s->data.power);
+				out.ryojouhen.horn = 0xFF; // Dedicated horn button, skip.
+				out.ryojouhen.hat = s->data.hatswitch & 0x0F;
+				out.ryojouhen.buttons = dct03_buttons(s->data.buttons);
+				break;
+			default:
+				Console.Error("Unhandled TrainController USB_TOKEN_IN pid=%d ep=%u type=%u", p->pid, p->ep->nr, s->type);
+				p->status = USB_RET_IOERROR;
+				return;
+		}
+
+		usb_packet_copy(p, &out, sizeof(out));
+	}
+
+	USBDevice* TrainDevice::CreateDevice(SettingsInterface& si, u32 port, u32 subtype) const
+	{
+		TrainDeviceState* s = new TrainDeviceState(port, static_cast<TrainDeviceTypes>(subtype));
+
+		s->desc.full = &s->desc_dev;
+
+		switch (subtype)
+		{
+			case TRAIN_TYPE2:
+				s->desc.str = dct01_desc_strings;
+				if (usb_desc_parse_dev(dct01_dev_descriptor, sizeof(dct01_dev_descriptor), s->desc, s->desc_dev) < 0)
+					goto fail;
+				break;
+			case TRAIN_SHINKANSEN:
+				s->desc.str = dct02_desc_strings;
+				if (usb_desc_parse_dev(dct02_dev_descriptor, sizeof(dct02_dev_descriptor), s->desc, s->desc_dev) < 0)
+					goto fail;
+				break;
+			case TRAIN_RYOJOUHEN:
+				s->desc.str = dct03_desc_strings;
+				if (usb_desc_parse_dev(dct03_dev_descriptor, sizeof(dct03_dev_descriptor), s->desc, s->desc_dev) < 0)
+					goto fail;
+				break;
+
+			default:
+				goto fail;
+		}
+
+		if (usb_desc_parse_config(taito_denshacon_config_descriptor, sizeof(taito_denshacon_config_descriptor), s->desc_dev) < 0)
+			goto fail;
+
+		s->dev.speed = USB_SPEED_FULL;
+		s->dev.klass.handle_attach = usb_desc_attach;
+		s->dev.klass.handle_reset = train_handle_reset;
+		s->dev.klass.handle_control = train_handle_control;
+		s->dev.klass.handle_data = train_handle_data;
+		s->dev.klass.unrealize = train_handle_destroy;
+		s->dev.klass.usb_desc = &s->desc;
+		s->dev.klass.product_desc = s->desc.str[2];
+
+		usb_desc_init(&s->dev);
+		usb_ep_init(&s->dev);
+		train_handle_reset(&s->dev);
+
+		return &s->dev;
+
+	fail:
+		train_handle_destroy(&s->dev);
+		return nullptr;
+	}
+
+} // namespace usb_pad
```

Comment:
newline

Reply:
the newline is missing again

Reply:
There's something about my Visual Studio configuration that seems to leak this back in. It's probably a line feed setting or something, but I don't usually use this dev environment so the settings are a bit non-obvious to me. I'll take a look at this.